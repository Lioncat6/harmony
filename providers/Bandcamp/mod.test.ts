import { describeProvider, makeProviderOptions } from '@/providers/test_spec.ts';
import { stubProviderLookups } from '@/providers/test_stubs.ts';
import { assert } from 'std/assert/assert.ts';
import { afterAll, describe } from '@std/testing/bdd';
import { assertSnapshot } from '@std/testing/snapshot';

import BandcampProvider from './mod.ts';

describe('Bandcamp provider', () => {
	const bc = new BandcampProvider(makeProviderOptions());
	const lookupStub = stubProviderLookups(bc);

	describeProvider(bc, {
		urls: [{
			description: 'album page',
			url: new URL('https://theuglykings.bandcamp.com/album/darkness-is-my-home'),
			id: { type: 'album', id: 'theuglykings/darkness-is-my-home' },
			isCanonical: true,
		}, {
			description: 'album page with tracking parameter',
			url: new URL('https://hiroshi-yoshimura.bandcamp.com/album/flora?from=discover_page'),
			id: { type: 'album', id: 'hiroshi-yoshimura/flora' },
		}, {
			description: 'standalone track page',
			url: new URL('https://zeug.bandcamp.com/track/yeltsa-kcir'),
			id: { type: 'track', id: 'zeug/yeltsa-kcir' },
			serializedId: 'zeug/track/yeltsa-kcir',
			isCanonical: true,
		}, {
			description: 'artist page/subdomain',
			url: new URL('https://taxikebab.bandcamp.com/'),
			id: { type: 'artist', id: 'taxikebab' },
			isCanonical: true,
		}, {
			description: 'artist /music page',
			url: new URL('https://theuglykings.bandcamp.com/music'),
			id: { type: 'artist', id: 'theuglykings' },
			isCanonical: false,
		}, {
			description: 'URL without subdomain',
			url: new URL('https://bandcamp.com/discover'),
			id: undefined,
		}],
		releaseLookup: [{
			description: 'label release with fixed price (which is not free despite minimum_price of 0.0)',
			release: 'thedarkthursday/and-it-was-a-burned-into-my-mind-yet-i-faltered-like-a-broken-record',
			assert: async (release, ctx) => {
				await assertSnapshot(ctx, release);
				const isFree = release.externalLinks.some((link) => link.types?.includes('free download'));
				assert(!isFree, 'Release should not be downloadable for free');
				const accountName = 'thedarkthursday';
				assert(
					release.labels?.some((label) => label.externalIds?.some(({ id }) => id === accountName)),
					'Bandcamp account should be linked to a label',
				);
				assert(
					!release.artists?.some((artist) => artist.externalIds?.some(({ id }) => id === accountName)),
					'Bandcamp account should not be linked to an artist',
				);
			},
		}],
	});

	afterAll(() => {
		lookupStub.restore();
	});
});
