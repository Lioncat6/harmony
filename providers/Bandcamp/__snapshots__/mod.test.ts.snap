export const snapshot = {};

snapshot[`Bandcamp provider > release lookup > label release with fixed price (which is not free despite minimum_price of 0.0) 1`] = `
{
  artists: [
    {
      creditedName: "i miss you sis",
      name: "i miss you sis",
    },
  ],
  availableIn: [
    "XW",
  ],
  credits: "i miss you sis

suika.love",
  externalLinks: [
    {
      types: [
        "paid download",
        "free streaming",
      ],
      url: "https://thedarkthursday.bandcamp.com/album/and-it-was-a-burned-into-my-mind-yet-i-faltered-like-a-broken-record",
    },
  ],
  gtin: undefined,
  images: [
    {
      comment: undefined,
      thumbUrl: "https://f4.bcbits.com/img/a1695876544_9.jpg",
      types: [
        "front",
      ],
      url: "https://f4.bcbits.com/img/a1695876544_0.jpg",
    },
  ],
  info: {
    messages: [],
    providers: [
      {
        apiUrl: undefined,
        id: "thedarkthursday/and-it-was-a-burned-into-my-mind-yet-i-faltered-like-a-broken-record",
        internalName: "bandcamp",
        lookup: {
          method: "id",
          value: "thedarkthursday/and-it-was-a-burned-into-my-mind-yet-i-faltered-like-a-broken-record",
        },
        name: "Bandcamp",
        url: "https://thedarkthursday.bandcamp.com/album/and-it-was-a-burned-into-my-mind-yet-i-faltered-like-a-broken-record",
      },
    ],
  },
  labels: [
    {
      externalIds: [
        {
          id: "thedarkthursday",
          provider: "bandcamp",
          type: "artist",
        },
      ],
      name: "The Dark Thursday",
    },
  ],
  media: [
    {
      format: "Digital Media",
      tracklist: [
        {
          artists: undefined,
          length: 81250,
          number: 1,
          title: "suomi lovely",
        },
        {
          artists: undefined,
          length: 72019.2,
          number: 2,
          title: "bittersweet desires to overdose on my prescriptions",
        },
        {
          artists: undefined,
          length: 60942.3,
          number: 3,
          title: "set fire to the space snow",
        },
      ],
    },
  ],
  packaging: "None",
  releaseDate: {
    day: 17,
    month: 7,
    year: 2019,
  },
  status: "Official",
  title: "and it was a burned into my mind! yet i faltered like a broken record",
  types: undefined,
}
`;
